### DISCUSSION TITLE
SSH2 (branch v3) : exec and timeout : how to stop the command ?

### DISCUSSION THREAD

**Discussion by TiTidom-RC** [phpseclib/phpseclib#2046] [Q&A] [answered]
Hi,

I have a problem with SSH2 exec() and the timeout, and i don't know how to solve it :(

When the timeout is reached, the ssh2->isTimeout() is 'true' = all is good here.

but after timeout is reached, i want to stop the command on the ssh connection, and here is my problem : i don't find any way to stop the command properly.

an example to illustrate my problem : 

here is the command i'm launching with $ssh2->exec() : 

`while true; do echo $(( $(tail -n 1 /tmp/testfilessh.txt) + 1 )) >> /tmp/testfilessh.txt; sleep 1; done; echo 'ok'`

it's an infinite loop, but the $ssh2 session has a timeout defined on 30 seconds.

After 30 sec, the timeout is well raised and my php code continue, i'm testing the $ssh2-isTimeout() and if detected i want to stop the infinite loop (the command) on the remote ssh session, but i don't find how to do that :(

i've tried with `$ssh2->reset()` as described in the doc, but it does nothing (perhaps it's only for read() command ?)
i've tried with $ssh2->write("\x03") but it does nothing either.

So, how to stop the command executed with $ssh2->exec() ?

Thank's for your help.

**Comment by terrafrost** [comment, marked as answer]
That's more of a Linux question than it is a phpseclib question. Like you could even kill the phpseclib process and, at least in my testing, the `while true` loop continues to run.

When I did `ps aux` I saw `sleep 1`. When I tried to kill that pid that pid had already ended and another one had started up. Per https://askubuntu.com/a/1033882/180177 `ps fjx` will do the trick. When I ran that command I got this as the output:

```
      1 1225658 1225658 1225658 ?             -1 Ss       0   0:00 bash -c while true; do echo $(( $(tail -n 1 /tmp/testfilessh.txt) + 1 )) >> /tmp/testfiles
1225658 1225693 1225658 1225658 ?             -1 S        0   0:00  \_ sleep 1
```
In this case 1225658 is the pid so I just did `kill 1225658`. You could do all that with an interactive shell but it'd be tricky. If you just wanted to kill the command with another `$ssh->exec()` (without a PTY) this worked for me:

```
kill $(ps fjx | grep testfilessh.txt | grep -v grep | awk '{print $2}')
```
That said, it looks like if an `exec()` ends because of a timeout that you can't do any more `exec()` commands. In phpseclib 3.0.37 you could, altho in that one your `while true` never timed out at all whereas in the current version of phpseclib it does. Now you have to do `$ssh->reset()` before doing a subsequent `$ssh->exec()`.

`$ssh2->write("\x03")` would only work on an interactive channel or on a PTY.

**Reply by TiTidom-RC** [reply]
Hi, 

Thank's for your reply. 

Yes it's more a Linux question, but i thought that an ssh "reset" (via the lib) would kill the background process linked to this session, but you are right, this is not the case.

And thank you for the information about $ssh2->write("\x03") available only in an interfactive channel.

Have a nice day.
